### ISSUE TITLE
Query metrics in Grafana aren't aligned with the actual metrics in logs

### ISSUE THREAD

**Issue by thinline72** [milvus-io/milvus#4645]
**Describe the bug**
I've been doing some load testing with different indexes, dataset sizes and QPS (up to 2000). And I've noticed that my system starts to perform slow yet (default) Milvus metrics looks good in Grafana, so I wasn't sure where the bottleneck is. 

When I checked Milvus logs it turned that it's in Milvus. Please, see attached screenshots of grafana metrics and logs sample from the same run.

![image](https://user-images.githubusercontent.com/8447843/105407544-c266e280-5c25-11eb-92a5-e7ce42dda5f2.png)
![image](https://user-images.githubusercontent.com/8447843/105407568-cabf1d80-5c25-11eb-98e5-f86ab701ef68.png)

**Environment details**
- Milvus is running on GCP n2-standart-16 (16 CPU, 64RAM)
- kubernetes
- Milvus version v0.10.2 (0.10.2-cpu-d081520-8a2393 image tag)
- Milvus configuration: default config

Just in case: queries are sent one-by-one via HTTP from the same cluster. Had dependency issues to use GRPC that hopefully will be sorted out soon.

**Comment by fishpenguin** [comment]
The time of grafana screenshot is inconsistent to the time of the log. So I'm not sure if the happened at the same time.
What's more, the Query elapsed time in grafana is used to record the quantile of 0.8, 0.9 and 0.95 of query time. but not the realtime time.

**Comment by thinline72** [comment]
@fishpenguin These are results from the same run. Time is consistent, but timezones (hours) are not. Not sure why, but it seems like timezone in Milvus logs is 8h ahead of mine timezone (UTC+01:00). And all of these are running in the cloud somewhere in US timezones :)

But it's from the same run. As I said I'm doing load testing so I'm monitoring it right away.

I know it's quantiles, so that's why it's even more weird as 0.95 should have captured that. For example, these are the Grafana metrics for the same run for another service that encodes query into vector and queries Milvus. As you can see p95 is much more consistent to the Milvus logs rather than to the Milvus Grafana:
![image](https://user-images.githubusercontent.com/8447843/105475430-486f4180-5c97-11eb-9930-7e5a3825d7fc.png)
![image](https://user-images.githubusercontent.com/8447843/105475573-68066a00-5c97-11eb-807c-c50558fed28c.png)

**Comment by thinline72** [comment]
I'll run a high load test right now that would produce very high metrics in the Milvus logs yet low in the Grafana and share results here.

**Comment by thinline72** [comment]
Queries: 2000 RPS, 4 min including 1min ramp
Index: ~75k vectors, 128dim, HNSW index (it'll be even worse with FLAT) 

![image](https://user-images.githubusercontent.com/8447843/105478103-a6e9ef00-5c9a-11eb-9bd1-5f46104058b4.png)
![image](https://user-images.githubusercontent.com/8447843/105477678-16abaa00-5c9a-11eb-8fb9-6c72ea4c2d41.png)

[The logs for this test run](https://drive.google.com/file/d/19hHom70t-jwbjpZk4jJCSVFYCBviYOMJ/view?usp=sharing) (30MB+ zip file, so using Google Drive).

**Comment by fishpenguin** [comment]
I checked the query elapsed time in milvus, and found that this metric records the query time in milvus db engine, which is as "Engine query totally cost" in logs, but not the "Total cost" time.
So it seemed weird that there was so much time gap between "Total time" and "Engine query totally cost". I will do some tests and check the problem.

**Comment by thinline72** [comment]
Sounds great! Please, let me know if I can help somehow. As I mentioned in the first message, I use HTTP REST API and vectors are send in parallel by one (not in batch).

BTW, is there any metric that logs "Total time" into Grafana?

**Comment by thinline72** [comment]
I'm not 100% sure if it's related to the current issue, but I was trying to do load testing with GPU enabled. I see GPU was used for index building etc. + I set gpuSearchThreshold to the 500. But when I ran the load test on 1000 QPS it doesn't utilize the GPU for querying. 

I've checked the logs and `nq` is very low (~1-11) etc. I thought Milvus accumulates the incoming query vectors into batches, no? (as mentioned above in my case Milvus receives one vector per each request).

Just in case attaching [logs](https://drive.google.com/file/d/12T61blC3x3tFbLao9L3ROaUbyj7Wqxz1/view?usp=sharing) and grafana metrics:
![image](https://user-images.githubusercontent.com/8447843/105733905-f003a280-5f29-11eb-9209-7bd0c34f2911.png)

**Comment by fishpenguin** [comment]
>Sounds great! Please, let me know if I can help somehow. As I mentioned in the first message, I use HTTP REST API and vectors are send in parallel by one (not in batch).
In milvus, these http request will be put into a queue and do sequentially.

>BTW, is there any metric that logs "Total time" into Grafana?
Currently, there is no metric that logs "Total time" into grafana. If you really need this metric, you can register a issue, and we can consider to add it in later milvus version.

**Comment by fishpenguin** [comment]
>I've checked the logs and nq is very low (~1-11) etc. I thought Milvus accumulates the incoming query vectors into batches, no? (as mentioned above in my case Milvus receives one vector per each request).

I checked the log and found that the query request had been combined, but not all combined. Combined nq are all less than 500(gpuSearchThreshold), so gpu didn't be used to query.
The searchCombine only happens when current accumulat nq or current request nq are less than 64, for example, when there comes 100 query requests with nq=1, searchCombine will combine them to 2 query requests with nq=64 and nq=36.

**Comment by thinline72** [comment]
@fishpenguin is there a way to configure that `nq` value? It looks like atm in order to efficiently utilize GPU I'll have to implement such batching mechanisms on my side. But I think it'd be nice to have in Milvus as it already has the batching functionality. What do you think?

> If you really need this metric, you can register a issue, and we can consider to add it in later milvus version.
well, I guess it makes sense to log the total time that Milvus spent on queries. Otherwise metrics in the Grafana are misleading. I spent a lot of time trying to figure out where the bottleneck is in the system before actually checking the Milvus logs. So I think it might help others to see the whole picture, right?

Please, let me know if both these things make sense to you. I'll be happy to create a dedicated request issues.

**Comment by fishpenguin** [comment]
>@fishpenguin is there a way to configure that nq accumulation value? It looks like atm in order to efficiently utilize GPU I'll have to implement query batching mechanisms on my side. But I think it'd be nice to have it in Milvus as it already has such batching functionality? What do you think?

The nq accumulation value is hardcode in milvus, if you want to change this, you have to change the code. We use 64 because we only test for accumulation value of 64, so if the value changed to be larger, we can't be sure that if there would be problems. However, if you really want to use gpu, I still suggest you to implement query batching on your side or set lower gpuSearchThreshold(though I don't suggest this way).

**Comment by fishpenguin** [comment]
> well, I guess it makes sense to log the total time that Milvus spends on queries. Otherwise metrics in the Grafana are misleading. I spent a lot of time trying to figure out where the bottleneck is in the system before actually checking the Milvus logs. So I think it might help others to see the whole picture, right?

I think this make sense, so I will add this metric in milvus-0.10.6 based on your issue.

**Comment by thinline72** [comment]
Sounds very reasonable, thank you! Please, let me know if you want me to create additional ticket for the metric request.

BTW, were you able to figure out why there is such huge difference between engine cost and total time?

**Comment by fishpenguin** [comment]
It seems that the extra time cost in total time is spend on json parse period. Milvus http server will receive http request body as json string, and get vectors from these string, which has a low performance. 
If total time is import for you, I suggest that you use grpc server but not restful api.
